pkg/netns/netns_linux.go

```
@@ -0,0 +1,213 @@
+func GetNetNSCookie() (uint64, error) {
+	s, err := unix.Socket(unix.AF_INET, unix.SOCK_STREAM, 0)
+	if err != nil {
+		return 0, err
+	}
+	defer unix.Close(s)
+
+	cookie, err := unix.GetsockoptUint64(s, unix.SOL_SOCKET, SO_NETNS_COOKIE)
+	if err != nil {
+		return 0, err
+	}
+
+	return cookie, nil
+}
+
+// FD returns the underlying file descriptor of the netNS
+// Note that FDs should not be used for comparing or testing equality, for that
+// see `equals`
+func (h *NetNS) FD() int {
+	return h.fd
+}
+
+// Close closes the handle to the network namespace. Note that this does not guarantee
+// the network namespace itself will be destroyed or closed, which will happen when all
+// references to it (including the one being closed here) are gone.
+func (h *NetNS) Close() error {
+	if err := unix.Close(h.fd); err != nil {
+		return err
+	}
+	h.fd = -1
+	return nil
+}
+
+// Do runs the provided func in the netns without changing the calling thread's netns.
+//
+// The code in f and any code called by f must NOT call [runtime.LockOSThread],
+// as this could leave the goroutine created by Do permanently pinned to an OS thread.
+func (h *NetNS) Do(f func() error) error {
+	g := new(errgroup.Group)
+
+	// Start the func in a new goroutine and lock it to an exclusive thread. This ensures that if
+	// execution of the goroutine fails unexpectedly before we call UnlockOSThread, the go runtime
+	// will ensure the underlying OS thread is disposed of, rather than reused in a potentially
+	// undefined state.
+	//
+	// See also: https://pkg.go.dev/runtime#UnlockOSThread
+	g.Go(func() error {
+		// save a handle to current network namespace
+		// and lock the thread - this will force this goroutine to be
+		// scheduled in a different thread, so we should be safe to mutate the netns
+		restore, err := saveRestoreCurrentThreadNetNS()
+		if err != nil {
+			return err
+		}
+		// unconditionally restore the snapshotted netns on termination, no matter what
+		defer restore()
```

Could we remove the use of `defer` here? It makes it slightly easier to reason about and verify we're doing things in the right order:

```go
restoreUnlock, err := lockOSThreadNetNS()
...
unix.Setns()
ferr := f()
if err := restoreUnlock(); err != nil {
    return err
}
return ferr
```

We also need to return an error from `restoreUnlock` and make sure to never swallow it. If something goes wrong here, this must be passed back to the caller and logged.

---

Yep, removed, as per https://github.com/cilium/cilium/pull/29993#discussion_r1471754515